src/utilutxocommit.cpp

```diff
@@ -0,0 +1,124 @@
+#include "utilutxocommit.h"
+#include "coins.h"
+#include "primitives/transaction.h"
+#include "utxocommit.h"
+#include "util.h"
+
+#include <atomic>
+#include <chrono>
+#include <future>
+#include <mutex>
+#include <stdexcept>
+
+namespace {
+const size_t BUFFER_ENTRIES_PER_WORKER = 5000;
+
+class CircularBuffer {
+private:
+    size_t head = 0;
+    size_t tail = 0;
+    const size_t size = 0;
+    std::vector<std::unique_ptr<std::pair<COutPoint, Coin>>> buffer;
+public:
+    std::mutex cs;
+
+    CircularBuffer(size_t sizeIn) : size(sizeIn) {
+        buffer.resize(size);
+        for (auto& b : buffer)
+            b.reset(new std::pair<COutPoint, Coin>{});
+    }
+    bool IsFull() {
+        return ((head + 1) % size) == tail;
+    }
+    bool IsEmpty() {
+        return head == tail;
+    }
+    void Put(std::unique_ptr<std::pair<COutPoint, Coin> >& i) {
+        assert(!IsFull());
+        buffer[head].swap(i);
+        head = (head + 1) % size;
+    }
+    void Get(std::unique_ptr<std::pair<COutPoint, Coin> >& i) {
+        assert(!IsEmpty());
+        buffer[tail].swap(i);
+        tail = (tail + 1) % size;
+    }
+};
+} // ns anon
+
+CUtxoCommit BuildUtxoCommit(CCoinsViewCursor* pcursor, size_t numworkers)
+{
+    if (numworkers == 0)
+        throw std::invalid_argument("numworkers == 0");
+
+    CircularBuffer buffer(numworkers * BUFFER_ENTRIES_PER_WORKER);
+    std::atomic<bool> eof(false);
+
+    // spin up workers
+    auto worker = [&buffer, &eof]() {
+        CUtxoCommit commitment;
+        auto utxo = std::make_unique<std::pair<COutPoint, Coin>>();
```

Comment:
make_unique seems to be the first C++14 feature in the codebase. Not sure if intentional, but maybe just stick to std::unqiue_ptr constructor?